python/dgl/sampling/negative.py

```diff
@@ -0,0 +1,72 @@
+"""Negative sampling APIs"""
+
+from .._ffi.function import _init_api
+from .. import backend as F
+
+__all__ = [
+    'global_uniform_negative_sampling']
+
+def global_uniform_negative_sampling(
+        g, num_samples, exclude_self_loops=True, unique=True, etype=None, num_trials=3,
+        redundancy=1.3):
+    """Performs negative sampling, which generate source-destination pairs such that
+    edges with the given type do not exist.
+
+    Specifically, this function takes in an edge type and a number of samples.  It
+    returns two tensors ``src`` and ``dst``, the former in the range of ``[0, num_src)``
+    and the latter in the range of ``[0, num_dst)``, where ``num_src`` and ``num_dst``
+    represents the number of nodes with the source and destination node type respectively.
+    It guarantees that no edge will exist between the corresponding pairs of ``src``
+    with the source node type and ``dst`` with the destination node type.
+
+    .. note::
+
+       This function uses rejection sampling, and may not always return the same number
+       of negative examples as the given :attr:`num_samples`.  This is more likely to
+       happen when a graph is so small or dense that not many unique negative examples exist.
+
+    Parameters
+    ----------
+    g : DGLGraph
+        The graph.
+    num_samples : int
+        The number of negative examples to generate.
+    exclude_self_loops : bool, optional
+        Whether to exclude self-loops from the negative examples.  Only impacts the
+        edge types whose source and destination node types are the same.
+
+        Default: True.
```

Comment:
Could be.  Say you compute the score with a dot product between the representation of two entities, a self-loop as a negative example will essentially gear the representation's magnitude towards 0, which might be undesirable.  You could argue that this is a regularization technique, but I would rather leave this as an option just in case people would like to turn this off.

@zjost you could also add your thoughts on this.

Reply:
I know that PyG excludes self-loops from their negative samples, but I haven't invested energy into understanding why.  Intuitively, it seems strange that you would use a self-loop as a negative edge sample, but I haven't conducted or seen any studies of this question.

Reply:
Self loops should be excluded since if you add self loop as a negative edge the optimum solution for the optimization problem is a vector of all zero. I would remove self-loop.